agent/lib/kontena/websocket/client.rb

```diff
@@ -0,0 +1,767 @@
+  #
+  # @param message [String, Array<Integer>]
+  # @raise [ArgumentError] invalid type
+  # @raise [RuntimeError] unable to send (socket closed?)
+  def send(message)
+    case message
+    when String
+      with_driver do |driver|
+        fail unless driver.text(message)
+      end
+    when Array
+      with_driver do |driver|
+        fail unless driver.binary(message)
+      end
+    else
+      raise ArgumentError, "Invalid type: #{message.class}"
+    end
+  end
+
+  # Register pong handler.
+  # Called from the #read thread every ping_interval after received pong.
+  #
+  # The ping interval should be longer than the ping timeout.
+  # If new pings are sent before old pings get any response, then the older pings do not yield on pong.
+  #
+  # @yield [delay] received pong
+  # @yieldparam delay [Float] ping-pong delay in seconds
+  def on_pong(&block)
+    @on_pong = block
+  end
+
+  # Send ping message.
+  #
+  # This is intended to be automatically called from #run per @ping_interval.
+  # Calling it from the #run callbacks also works, but calling it from a different thread
+  # while #run is blocked on read() will ignore the @ping_timeout, unless the server happens
+  # to send something else.
+  #
+  # @raise [RuntimeError] not connected
+  def ping
+    with_driver do |driver|
+      # must be called from #read loop to use the right read timeout
+      ping_at = pinging!
+
+      debug "pinging at #{ping_at}"
+
+      fail unless driver.ping(ping_at.utc.strftime(PING_STRFTIME)) do
+        debug "pong for #{ping_at}"
+
+        # resolve ping timeout, unless this pong is late and we already sent a new one
+        # called with @mutex held
+        if ping_at == @ping_at
+          ping_delay = pinged!
+
+          debug "ping-pong at #{ping_at} in #{ping_delay}s"
+
+          # queue to call block without @mutex held
+          enqueue { @on_pong.call(ping_delay) } if @on_pong
+        end
+      end
+
+    end
+  end
+
+  # Waiting for pong from ping
+  #
+  # @return [Boolean]
+  def pinging?
+    !!@pinging
+  end
+
+  # Measured ping-pong delay from previous ping
+  #
+  # nil if not pinged yet
+  #
+  # @return [Float, nil]
+  def ping_delay
+    @ping_delay
+  end
+
+  # Send close frame. Waits for server to send back close frame, and then allows #run to return.
+  # Imposes a close timeout when called from #run blocks (run/on_message/on_pong do ...). If called from
+  # a different thread, then #run should eventually return, once either:
+  # * server sends close frame
+  # * server sends any other frame after the close timeout expires
+  # * the ping interval expires
+  #
+  # XXX: prevent #send after close?
+  #
+  # @param close [Integer]
+  # @param reason [String]
+  def close(code = 1000, reason = nil)
+    debug "close"
+
+    with_driver do |driver|
+      fail unless driver.close(reason, code) # swapped argument order
+
+      closing!
+    end
+  end
+
+#protected XXX: called by specs
```

Comment:
Reverted the `send_message` vs protected `send` specs, this is now back to `#protected XXX` which is TODO on refactoring those protected methods called by specs out to separate classes... most of that is the TCP/SSL connection stuff, which is entirely independent of the websocket client.